### PR TITLE
My solutions to testing exercises.

### PR DESCRIPTION
**Pull request by akam**

### CODE REVIEW

**Comment by mmmaaatttttt** [comment]
Nice work on the testing assignment. If you want to push yourself, here are some questions to think about:

- what if you pass in 0 for the number in `expand`?
- Do you think `expand` should mutate the original array? Regardless of what you decide, could you write some tests to ensure that `expand` behaves as you expect?

**Comment by aykaypkt** [comment]
- what if you pass in 0 for the number in expand?

Initially when I tested it, it returned the original array, but logically it should copy it 0 times. now it returns an empty array.

- Do you think expand should mutate the original array? Regardless of what you decide, could you write some tests to ensure that expand behaves as you expect?

Initially I thought it would mutate the original array as I looked at the code and it assigns the new concat() to arr. But when testing I found it does not. I have made an educated guess that the array I pass in, is in the scope of the function so when the function ends, anything I have done to the function has been deleted and the array is returned as another variable and not the original array.

File: call_apply_bind_exercise/callApplyBind.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+function invokeMax(fn, num){
+	var counter = 0;
+	var times = num;
+	return function(...inner){
+		if(counter < times){
+			counter++;
+			return fn.apply(fn, inner);
+		} else {
+			return "Maxed Out!";
```
Comment (elie) [inline call_apply_bind_exercise/callApplyBind.js]: no need for an else!

File: call_apply_bind_exercise/callApplyBind.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	var sum = 0;
+	for(var val of nums){
+		if(val % 2 === 0){
+			sum += val;
+		}
+	}
+	return sum
+}
+
+function arrayFrom(alo){
```
Comment (elie) [inline call_apply_bind_exercise/callApplyBind.js]: make this shorter :)

File: constructor_functions_exercise/constructors.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+function Parent(firstName, lastName, favoriteColor, favoriteFood){
+    this.firstName = firstName;
+    this.lastName = lastName;
+    this.favoriteColor = favoriteColor;
+    this.favoriteFood = favoriteFood;
+}
+
+function Child(firstName, lastName, favoriteColor, favoriteFood){
+	Parent.call(Child, firstName, lastName, favoriteColor, favoriteFood);
```
Comment (elie) [inline constructor_functions_exercise/constructors.js]: apply?

File: constructor_functions_exercise/readme.md [hunk 4]
```diff
@@ -2,19 +2,61 @@
...
+
+**3. What does the keyword `this` refer to inside of a constructor function? **
+
+The key word `this` depends on the context given. If without the `new` word. `this` will refer to the closest object which might be the global window. When `this` is called with the `new` key word, it will refer to the object created by the `new` key word.
+
+**4. What is a `class`? What is an `instance`?**
+
+A Class is a blueprint which instances make use of to make new instances of the class. 
+
+In object-orientated programming. A Class would be the blueprints for the object and instances use these Classes to make objects. JavaScript tries to mimic this using Constructors for Classes
```
Comment (elie) [inline constructor_functions_exercise/readme.md]: spelling

File: lodash_exercise/lodash.js [hunk 5]
```diff
@@ -1,103 +1,325 @@
...
-
+function takeRight(arr){
+	if(arguments.length === 1) return arr.slice(-1);
+	if(arguments[1] === 0) return [];
+	return arr.slice(-arguments[1]);
 }
 
 function takeWhile(){
-
+	printf('Lol');
```
Comment (elie) [inline lodash_exercise/lodash.js]: printf???

File: lodash_exercise/lodash.js [hunk 6]
```diff
@@ -1,103 +1,325 @@
...
+		} else {
+			newArr.push(arr[i]);
+		}
+	}
+	return newArr;
 }
 
-function flattenDeep(){
-
+function flattenDeep(arr){
```
Comment (elie) [inline lodash_exercise/lodash.js]: how can we do this without helper method recursion?

File: recursion_exercise/recursion.js [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+		return index;
+	} else {
+		return found;
+	}
+}
+
+
+
+
+function binarySearch(array, key){
```
Comment (elie) [inline recursion_exercise/recursion.js]: try this without a helper

File: lodash_exercise/lodash.js (hunk 6 above)
Comment (aykaypkt) [inline lodash_exercise/lodash.js]: This one was hard. But eventually I got it..

Comment (aykaypkt) [inline lodash_exercise/lodash.js]: This one was hard but eventually I got it...

File: intermediate_oop/tic_tac_toe/js/board.js [hunk 8]
```diff
@@ -0,0 +1,79 @@
+function GameBoard() {
+  this.counter = 0;
+  this.turn = "X";
+  this.check = [1,2,3,4,5,6,7,8,9];
+  this.gameOn = true;
+}
+
+GameBoard.prototype.gameState = function(){
+	return this.gameOn;
```
Comment (elie) [inline intermediate_oop/tic_tac_toe/js/board.js]: watch your indentation

File: intermediate_oop/tic_tac_toe/js/board.js [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+GameBoard.prototype.getTurn = function() {
+  return this.turn;
+}
+
+GameBoard.prototype.changeTurn = function() {
+  if (this.turn === "X") {
```
Comment (elie) [inline intermediate_oop/tic_tac_toe/js/board.js]: a ternary could be a nice refactor

File: intermediate_oop/tic_tac_toe/js/board.js [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+GameBoard.prototype.add = function(number) {
+  this.check[number] = this.turn;
+}
+
+GameBoard.prototype.result = function(check, counter) {
+
+  if (this.check[0] === this.check[3] && this.check[3] === this.check[6]){
+  	document.getElementById("message").innerText = (this.check[0] + " wins");
+  	this.gameOn = false;
```
Comment (elie) [inline intermediate_oop/tic_tac_toe/js/board.js]: do we need to repeat `this.gameOn = false` every time?

File: intermediate_oop/tic_tac_toe/js/boardview.js [hunk 11]
```diff
@@ -0,0 +1,23 @@
+function BoardView(game) {
+  this.container = document.getElementById("board");
+  this.game = game;
+  $(this.container).on("click", ".square", this.draw.bind(this));
+  $("#new-game").on("click", this.reset.bind(this));
```
Comment (elie) [inline intermediate_oop/tic_tac_toe/js/boardview.js]: Don't mix jQuery with vanilla JS - use one or the other!

File: intermediate_oop/tic_tac_toe/js/boardview.js [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+  	this.game.upCounter();
+  	this.game.changeTurn();
+  	this.game.result(this.game.check, this.game.counter);
+  }
+
+}
+
+BoardView.prototype.reset = function() {
+  this.game.reset();
+  $(".square").text("");
```
Comment (elie) [inline intermediate_oop/tic_tac_toe/js/boardview.js]: Use vanilla JS here!

File: intermediate_oop/tic_tac_toe/js/board.js [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+  }
+  if (this.counter === 9) {
+    document.getElementById("message").innerText = ("It's a Tie!");
+    this.counter = 0;
+    this.gameOn = false;
+  }
+}
+
+GameBoard.prototype.upCounter = function(){
+	this.counter++;
```
Comment (elie) [inline intermediate_oop/tic_tac_toe/js/board.js]: watch your indentation - keep it consistent

Comment (aykaypkt) [inline intermediate_oop/tic_tac_toe/js/board.js]: Made changes to all the comments :)

**Pull request closed without merging**